messages_example_test.go

```diff
@@ -0,0 +1,89 @@
+// +build go1.10
+
+package mssql_test
+
+import (
+	"context"
+	"database/sql"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+
+	mssql "github.com/denisenkom/go-mssqldb"
+	"github.com/golang-sql/sqlexp"
+)
+
+const (
+	msgQuery = `select name from sys.tables
+PRINT N'This is a message'
+select 199
+RAISERROR (N'Testing!' , 11, 1)
+select 300
+`
+)
+
+// This example shows the usage of sqlexp/Messages
+func ExampleRows_usingmessages() {
+	flag.Parse()
+
+	if *debug {
+		fmt.Printf(" password:%s\n", *password)
+		fmt.Printf(" port:%d\n", *port)
+		fmt.Printf(" server:%s\n", *server)
+		fmt.Printf(" user:%s\n", *user)
+	}
+
+	connString := makeConnURL().String()
+	if *debug {
+		fmt.Printf(" connString:%s\n", connString)
+	}
+
+	// Create a new connector object by calling NewConnector
+	connector, err := mssql.NewConnector(connString)
+	if err != nil {
+		log.Println(err)
+		return
+	}
+
+	// Pass connector to sql.OpenDB to get a sql.DB object
+	db := sql.OpenDB(connector)
+	defer db.Close()
+	retmsg := &sqlexp.ReturnMessage{}
+	ctx := context.Background()
+	rows, err := db.QueryContext(ctx, msgQuery, retmsg)
+	if err != nil {
+		log.Fatalf("QueryContext failed: %v", err)
+	}
+	active := true
+	for active {
+		msg := retmsg.Message(ctx)
+		switch m := msg.(type) {
+		case sqlexp.MsgNotice:
+			fmt.Println(m.Message)
+		case sqlexp.MsgNext:
+			inresult := true
+			for inresult {
+				inresult = rows.Next()
+				cols, err := rows.Columns()
+				if err != nil {
+					log.Fatalf("Columns failed: %v", err)
+				}
+				fmt.Println(cols)
+				if inresult {
+					var d interface{}
+					if err = rows.Scan(&d); err != nil {
+						fmt.Println(d)
+					}
+				}
+			}
+		case sqlexp.MsgNextResultSet:
+			active = rows.NextResultSet()
+		case sqlexp.MsgError:
+			fmt.Fprintln(os.Stderr, m.Error)
+		case sqlexp.MsgRowsAffected:
+			fmt.Println("Rows affected:", m.Count)
+		}
+	}
+
```

Comment:
This test compiles but does not run, because it doesn't have an "Output:" comment. That could be intentional, because this test requires a working SQL Server and the output of this test is not deterministic (see my earlier comment about selecting name from sys.tables). Was it intentional?

Whether or not preventing this test from running was intentional, I think it would be a good to idea to make this test runnable. Doing so revealed a number of problems in my testing.

To make this test runnable, you will need to:
1. Add an "Output:" comment here with the expected stdout results.
2. Add code to skip this test if the connection parameters aren't specified. This is another instance where using the SQLSERVER_DSN environment variable to convey connection information is more convenient than using the flags. The testConnParams function that extracts connection information from SQLSERVER_DSN already includes a t.Skip call to skip the test when connection information is absent.
3. Remove the printing of the command line arguments and the connection string. They will vary from test to test, and thus will cause the test to fail.
4. Replace the log calls with fmt.Printf calls. That will capture errors in stdout where they can be both seen and detected by the example test pass logic.
5. Replace the stderr output with stdout output. Again all output needs to go the stdout so that it can be captured and detected by example test pass logic.

I imagine you used log and stderr because you were in the mindset that an example should show how those types of messages would likely be handled differently in an actual application. But I think the benefits of having a runnable example test carry more weight. You can add comments to those lines suggesting that in a real application doing something other than printing to stdout would be more appropriate, but I think most people already understand that.